### PR TITLE
[WIP] Refactor of dbcsr_api.F grid initialization, timers, and mp_env

### PR DESCRIPTION
**Pull request by dithillobothrium**

### CODE REVIEW

File: src/core/dbcsr_lib.F [hunk 1]
```diff
@@ -39,25 +64,99 @@ MODULE dbcsr_lib
    CHARACTER(len=*), PARAMETER, PRIVATE :: moduleN = 'dbcsr_lib'
 
    PUBLIC :: dbcsr_init_lib, dbcsr_finalize_lib, dbcsr_clear_mempools
+   PUBLIC :: dbcsr_init_timer,dbcsr_kill_timer, dbcsr_print_timer
```
Comment (alazzaro) [inline src/core/dbcsr_lib.F]: kill? I would propose remove instead. BTW, why these functions are public? They should be private...

Comment (dithillobothrium) [reply src/core/dbcsr_lib.F]: Ok, I will make them private. Kill sounds better. But remove maybe is not very correct, maybe `dbcsr_finalize_timer`

Comment (alazzaro) [reply src/core/dbcsr_lib.F]: Oh yes, finalize is better

**Comment by dithillobothrium** [comment]
I think one needs to pass in `dbcsr_init_lib` 2d communicator, because 2d communicator is passed in `dbcsr_finlaize_lib`. 2 different communicators in init and finalize is not obvious

**Comment by alazzaro** [comment]
Finalize use 1d. Then, since we set the comm on init, we can remove it from finalize.

**Comment by dithillobothrium** [comment]
> Finalize use 1d. Then, since we set the comm on init, we can remove it from finalize.

in some samples I saw 2d. May be it was wrong sample.

**Comment by alazzaro** [comment]
Thanks finalize requires the comm to print statistics, doesn't matter 1d or 2d. Look at the examples

**Comment by dithillobothrium** [comment]
Some error related to timers, but I dont understand

` *******************************************************************************
 *   ___                                                                       *
 *  /   \                                                                      *
 * [ABORT]                                                                     *
 *  \___/              list_timerenv_peek: list is not initialized.            *
 *    |                                                                        *
 *  O/|                                                                        *
 * /| |                                                                        *
 * / \                                                dbcsr_list_timerenv.F:27 *
 *******************************************************************************
`

**Comment by dithillobothrium** [comment]
now its done. Need to add hooks in the interface as a second part of the refactor.

**Comment by alazzaro** [comment]
test this please

**Comment by dithillobothrium** [comment]
I did not see this test yesterday. I will do.

**Comment by hfp** [comment]
This PR appears to require some changes in CP2K?

I am getting:

```text
/nfs/home2/hpabst/cp2k/src/f77_interface.F:312:39:

          CALL dbcsr_finalize_lib(group=default_para_env%group, output_unit=0)
                                       1
Error: Keyword argument ‘group’ at (1) is not in the procedure
/nfs/home2/hpabst/cp2k/src/f77_interface.F:277:30:

          CALL dbcsr_init_lib()
                              1
Error: There is no specific subroutine for the generic ‘dbcsr_init_lib’ at (1)
```

**Comment by alazzaro** [comment]
This is v2, it breaks the API (and other changes are expected).

**Comment by hfp** [comment]
Rebasing(?) such a PR is somewhat strange since all changes appeared somewhere in the past (Git history), but though the latest commit is still shown as from 26 days ago...

**Comment by dithillobothrium** [comment]
> Rebasing(?) such a PR is somewhat strange since all changes appeared somewhere in the past (Git history), but though the latest commit is still shown as from 26 days ago...

I see that, it is 3 days ago in dbcsr:develop and here too

**Comment by hfp** [comment]
Interesting, for me DBCSR's Git-home looks like the image below (lower right corner: "26 day ago").

grafik

**Comment by dithillobothrium** [comment]
> Interesting, for me DBCSR's Git-home looks like the image below (lower right corner: "26 day ago").
> 
Yes, it is the same for me. Strange...

**Comment by shoshijak** [comment]
Yes, this is correct behaviour and it is "my fault". 

I merged PR #172 today, though it contains some commits that I made quite a while ago. I just wanted to test these commits on different machines and some time passed between when I made these commits locally and when I merged them in a PR. 

So even though my commits were made locally at a date earlier than the ones in this PR ("26 days ago", as shown in github's interface), PR #172 was nevertheless "rebased and merged" after PR #171, which is why the commits of PR #172  appear more recently in the commit logs.

**Comment by hfp** [comment]
Is there an interest to get a CP2K PR to accomodate this PR/change?

**Comment by alazzaro** [comment]
Yes, but we are working on other issues for v2...

**Comment by hfp** [comment]
I have sent PR #176 which passed regtests (now fails on Daint after rebasing dbcsr/head).

Anyhow, this PR #171 (Refactor of dbcsr_api.F grid initialization, timers, and mp_env) leads to the following output when CP2K terminates (no matter which input/workload is given):

```text
 *******************************************************************************
 *   ___                                                                       *
 *  /   \                                                                      *
 * [ABORT]                                                                     *
 *  \___/              list_timerenv_peek: list is not initialized.            *
 *    |                                                                        *
 *  O/|                                                                        *
 * /| |                                                                        *
 * / \                                                dbcsr_list_timerenv.F:27 *
 *******************************************************************************
```

**Comment by alazzaro** [comment]
@hfp CP2K will come at the later stage...
For the reference, the problem is somehow related to the timeset/timestop used after finalize. Do you have the CP2K stacktrace?

**Comment by dithillobothrium** [comment]
now one needs to call dbcsr_finalize before the final MPI_Comm_free().
Probably I forgot to nullify callbacks in dbcsr_finalize_timer. Or something like that

**Comment by hfp** [comment]
I will look into this tomorrow - thank you all for the hints!

**Comment by dithillobothrium** [comment]
so, it should be
> now one needs to call dbcsr_finalize before the final MPI_Comm_free().
DBCSR should be fine internally

**Comment by hfp** [comment]
One comment: I cannot find the routine `dbcsr_finalize_timer` anywhere in the DBCSR's HEAD revision. Perhaps this was solved differently over the course of the PR or something is missing...

**Comment by alazzaro** [comment]
@hfp correct, please use the develop branch

**Comment by dithillobothrium** [comment]
it is built in dbcsr_finalize_lib

**Comment by hfp** [comment]
> please use the develop branch

Hmm, I should probably stop for today - I think I am on develop, but there is no `dbcsr_finalize_timer` anywhere in the source code. Though, since develop is the default branch eben a plain websearch should also show it: [https://github.com/cp2k/dbcsr/search?q=dbcsr_finalize_timer](https://github.com/cp2k/dbcsr/search?q=dbcsr_finalize_timer)

**Comment by alazzaro** [comment]
Sorry for the confusion, as Ilia said, that function is now inlined in dbcsr_finalize_lib...

**Pull request merged**